File: go/core/interfaces/leader.go
package interfaces

import "context"

// LeaderElector coordinates single-instance execution (D-30): only the elected leader
// runs the guarded work, so a job scheduled across N replicas fires once.
//
// Implementations: a single-process elector that is always the leader (dev/single-replica),
// or a distributed elector (a Postgres advisory lock, the sibling of clients/lock's
// postgres backend). Job code depends on this interface, never a concrete elector.
type LeaderElector interface {
	// IsLeader reports whether this instance currently holds leadership.
	IsLeader(ctx context.Context) (bool, error)
}

File: python/techai_webutils/src/techai_webutils/core/interfaces/leader.py
"""Leader-election interface for single-instance execution.

Mirrors Go's ``interfaces.LeaderElector``. Only the elected leader runs the guarded work,
so a job scheduled across N replicas fires once (D-30).
"""

from __future__ import annotations

from typing import Protocol, runtime_checkable


@runtime_checkable
class LeaderElector(Protocol):
    """Coordinates single-instance execution (D-30).

    Implementations: a single-process elector that is always the leader (dev/single-replica),
    or a distributed elector (a Postgres advisory lock, the sibling of ``clients/lock``'s
    postgres backend). Job code depends on this Protocol, never a concrete elector.
    """

    async def is_leader(self) -> bool:
        """Report whether this instance currently holds leadership."""
        ...
